File: backend/express/get-data.js
import { spawn } from 'child_process';
import Content from '../mongo-db/content.js';
import Query from '../mongo-db/query.js';

const getDefault = (key) => {
    const defaults = {
        dateStart: "1111-11-11",
        dateEnd: "9999-09-09",
        source: "ALL",
        sentiment: "ALL",
        model: "LOCAL"
    };
    return defaults[key];

}

const getData = async (req, res) => {
    // Step 1: Interpret the post req parameters
    let { query, dateStart, dateEnd, source, sentiment } = req.body;
    console.log(`Received req from client: \n\tquery: ${query}\n\tdateStart: ${dateStart}\n\tdateEnd: ${dateEnd}\n\tsource: ${source}\n\tsentiment: ${sentiment}`)
    if (!query) {
        return res.json({});
    }
    dateStart = dateStart || getDefault("dateStart");
    dateEnd = dateEnd || getDefault("dateEnd");
    source = source || getDefault("source");
    sentiment = sentiment || getDefault("sentiment");
    
    
    const dateStartObject = new Date(dateStart)
    const dateEndObject = new Date(dateEnd)
    // Step 2: Check if parameters are in database, or a superset is.
    const cachedQuery = await Query.findOne({
        keyword: query,
        $or: [
            { source: source },
            { source: "ALL" }
        ],
        datestart: { $lte: dateStartObject },
        dateend: { $gte: dateEndObject },
    }).populate('contentid');
    // Step 3: If query is already in database, return results that match from superset
    if (cachedQuery) {
        const filteredContent = cachedQuery.contentid.filter(content =>
            (source === "ALL" || content.source === source) &&
            content.date >= dateStartObject &&
            content.date <= dateEndObject
        );
        console.log(`Found ${filteredContent.length} matching of ${cachedQuery.contentid.length} relavent results in database`)
        const content = filteredContent.map(content => ({
            title: content.title,
            username: content.username,
            content_body: content.content_body,
            date: content.date.toISOString(),
            source: content.source,
            source_url: content.source_url,
            explicit: "FALSE",
            sentiment: "NEUTRAL"
        }));
        console.log(`Response of ${content.length} results sent to client\n`)
        return res.json(content);
    }

    console.log("Failed to find results in database")
    //Step 4: Fallthrough; Spawn python process to scrape data
    const args = [query, dateStart, dateEnd, source, sentiment];
    const pythonProcess = spawn('python3', ['python/get_data.py', JSON.stringify(args)]);

    let pythonOutput = '';
    let errorData = '';

    pythonProcess.stdout.on('data', (chunk) => {
        pythonOutput += chunk.toString();
    });

    pythonProcess.stderr.on('data', (chunk) => {
        errorData += chunk.toString();
    });

    pythonProcess.on('close', async (code) => {
        console.log(errorData.trim())
        if (code !== 0 || !pythonOutput) {
            console.log(`Error while executing python code: get_data.py with parameters "${args}"`)
            console.log(`System Time: ${new Date()}`)
            return res.status(500).send(`Server failed to gather data`);
        }
        try {
            const contentList = JSON.parse(pythonOutput.trim());

            // Step 5: Save content from python process in database
            const contentDocs = await Content.insertMany(contentList.map(item => ({
                title: item.title,
                username: item.username,
                content_body: item.content_body,
                date: new Date(item.date),
                source: source,
                source_url: item.source_url
            })));
            console.log(`Inserted ${contentList.length} new results to database`)
            const contentIds = contentDocs.map(doc => doc._id);
            // Step 6: If new data is a supeset of any queries, merge them all into 1 new query 
            const subsetQueries = await Query.find({
                keyword: query,
                datestart: { $gte: dateStartObject },
                dateend: { $lte: dateEndObject },
                ...(source !== "ALL" && { source: source }), // Add `source` condition only if it's not "ALL"
            });
            let mergedContentIds = new Set(contentIds);
            subsetQueries.forEach(subsetQuery => {
                subsetQuery.contentid.forEach(id => mergedContentIds.add(id.toString()));
            });
            const subsetQueryIds = subsetQueries.map(q => q._id);
            await Query.deleteMany({ _id: { $in: subsetQueryIds } });
            console.log(`Deleted ${subsetQueries.length} subset queries`)
            // Step 7: Save the new query in the database
            const newQuery = new Query({
                keyword: query,
                source: source,
                datestart: dateStartObject,
                dateend: dateEndObject,
                requestcount: 10,
                contentid: Array.from(mergedContentIds),
            });

            await newQuery.save();
            console.log(`Saved new query to database`)
            //Step 8: Return results to client
            console.log(`Response of ${contentList.length} results sent to client\n`)
            res.json(contentList);


        } catch (err) {
            console.error('Error saving content to DB:', err);
            res.status(500).send('Unknown Server Error');
        }
    });

}
export default getData;

File: backend/routes/get_data.py
from flask import request, jsonify
from datetime import datetime
from schemas.content_schema import Content
from schemas.query_schema import Query
from scraping.scrape_data import scrape_data
def get_default(key):
    current_year = datetime.now().year
    defaults = {
        "dateStart": f"{current_year}-01-01",
        "dateEnd": "9999-09-09",
        "source": "ALL",
        "sentiment": "ALL",
        "model": "LOCAL"
    }
    return defaults.get(key)

def get_data():
    data = request.get_json()
    
    query = data.get("query")
    if not query:
        return jsonify([])
    date_start = datetime.strptime(data.get("dateStart") or get_default("dateStart"), "%Y-%m-%d")
    date_end = datetime.strptime(data.get("dateEnd") or get_default("dateEnd"), "%Y-%m-%d")
    source = data.get("source", get_default("source"))
    sentiment = data.get("sentiment", get_default("sentiment"))
    print("\nReceived Request with Parameters:")
    print("\tQuery:",query)
    print("\tDate Start:",date_start)
    print("\tDate End:",date_end)
    print("\tSource:",source)
    print("\tSentiment:",sentiment)

    
    
    # Step 2: Check if query is in database
    cached_query = Query.objects(
        keyword=query,
        source__in=[source, "ALL"],
        datestart__lte=date_start,
        dateend__gte=date_end,
    ).first()
    
    if cached_query:
        filtered_content = [c for c in cached_query.contentid
                            if (source == "ALL" or c.source == source) and
                            date_start <= c.date <= date_end]
        print(f"Found {len(filtered_content)} matching of {len(cached_query.contentid)} relavent results in database")
        return jsonify([{
            "title": c.title,
            "username": c.username,
            "content_body": c.content_body,
            "date": c.date.strftime("%Y-%m-%d"),
            "source": c.source,
            "source_url": c.source_url,
        } for c in filtered_content])

    # Step 3: Fallthrough; Call data collection function
    print(f"No results in database, Starting data collection at {datetime.now()}")
    content_list = scrape_data(query, date_start, date_end, source, sentiment)
    print(f"Data collection finished at {datetime.now()}")

    # Step 4: Save new content to database
    print(f"Saving results to database...")
    content_docs = [Content(**item).save() for item in content_list]
    content_ids = [c.id for c in content_docs]
    
    # Step 5: Merge subset queries
    subset_queries = Query.objects(
        keyword=query,
        datestart__gte=date_start,
        dateend__lte=date_end,
        **({"source": source} if source != "ALL" else {})
    )
    print(f"Found {len(subset_queries)} related queries. Merging...")
    merged_content_ids = set(content_ids)
    for subset in subset_queries:
        merged_content_ids.update(subset.contentid)
    
    subset_queries.delete()
    
    # Step 6: Save new query
    new_query = Query(
        keyword=query,
        source=source,
        datestart=date_start,
        dateend=date_end,
        requestcount=10,
        contentid=list(merged_content_ids)
    )
    new_query.save()
    print(f"Finished saving results.")

    return jsonify(content_list)